sympy/matrices/tests/test_matrices.py

```diff
@@ -2075,6 +2086,16 @@ def test_cholesky_solve():
     b = A*x
     soln = A.cholesky_solve(b)
     assert soln == x
+    A = Matrix(((9, 3*I), (-3*I, 5)))
```

Comment:
This test and the test below it still pass in SymPy 1.1.1.

Reply:
There is probably no reason why the formulas developed for decomposition of symmetric matrices would not work also for non-real matrices (principle of analytic continuation). I think that the main reason why the hermitian symmetric version seems to be generally favoured for Cholesky decomposition is that the radicands of square roots are real and positive. SymPy can also handle the non-real case but that may lead to complicated nested root expressions.

Reply:
@jksuom do you think it will be valid to replace `if self.is_hermitian` with `if self.is_hermitian or self.is_symmetric()` in this pull request? c.f. https://github.com/sympy/sympy/issues/14672

Reply:
The same code does not work for both symmetric and hermitian matrices. The current implementation is only for hermitian matrices while the old one was only for symmetric matrices.

It would be possible to have both implementations. The symmetric one could be used (automatically, or on a hint) for real hermitian (= symmetric) matrices.

Reply:
If the matrix entries are all real, the matrix will already be detected as hermitian and the current code should apply (the `conjugate` calls will do nothing). The issue here is for symmetric (not Hermitian) non-real matrices. 

Is the algorithm difference just a question of calling `conjugate` in various places only in the Hermitian case? 

The benchmark that is slowed by this PR is `A_sym.cholesky_solve(b)` on 

```py
A_sym = Matrix(n, n, lambda i, j: Symbol('a{}{}'.format(*sorted((i, j)))))
b = Matrix(n, 1, lambda i, j: Symbol('b{}{}'.format(i, j)))
```

for `n = 3`, i.e., a general symbolic Cholesky solve for a 3 x 3 symmetric matrix. 

The answer before this pull request is much simpler and returns much faster. According to `simplify(A_sym*A_sym.cholesky_solve(b)) == b`, the old answer was correct. Assuming `simplify` isn't doing any invalid this shows that the old algorithm was correct for the general 3x3 case.

Reply:
> Is the algorithm difference just a question of calling conjugate in various places only in the Hermitian case?

This looks like the case. The main changes here were adding some `conjugate` calls and swapping `M.T` vs. `M.H`. Also is_positive checks for the diagonal were added. 

The PR description says that the old algorithm was wrong for complex symmetric matrices, though the above seems to show that it isn't at least for the 3x3 case. So that's what I'd like to clarify. If the old algorithm was incorrect, we should just leave it. Even if it's slower it's right now. But then we should look into why `simplify` seems to show that it is (or is more than 3x3 needed to show this?). If it was correct, we should add it back in, as the new output for symmetric nonreal matrices is much slower.

Reply:
The most time-consuming part, in a symbolic system, is `conjugate`. In a numeric system working with native complex numbers that should not take much time. Conjugation can be avoided in the code, if desired, by adding a keyword argument `real` and replacing `conjugate` by a local `conj` as follows.

    if real:
        conj = lambda x: x
    else:
        conj = conjugate

This will be essentially the old implementation when `real` is `True`. It can also be used for complex (symmetric) matrices but then the diagonal square root terms will be complex (and potentially complicated in a bigger matrix).

> the old algorithm was wrong for complex symmetric matrices

It was wrong for the decomposition specified by Cholesky: `A = L*L.H` because it gives `A = L*L.T` which is different for complex matrices. But it is not wrong if we really want the latter 'non-Cholesky decomposition'. It is possible to use either decomposition in `cholesky_solve` (since it can comply with both). One should probably test with bigger than 3x3 matrices to see the advantages and disadvantages of the two decompositions in `cholesky_solve` for complex matrices. (`cholesky_solve` could also have the keyword `real`.)

Reply:
Another question is whether we really want to use `cholesky_solve` (any version) for symbolic matrices as there is `LUsolve` that does not introduce square roots.

Reply:
One reason to use Cholesky decomp over LU decomp is that it should be faster (half the speed) when the matrix is symmetric and positive (semi) definite. This would be very nice to use in the mechanics package because we almost always have this special case for the important linear systems that need solving. But, our cholesky implementation has never been faster than LU decomp and this PR makes it quite worse. Most people would probably be expecting a speed up if they make use of the Cholesky algorithm.